File: leetcode/golang/5.go
package leetcode

func getLongestPalindrome(s string, l int, r int) (string, int) {
	length := len(s)
	for l-1 >= 0 && r+1 < length {
		if s[l-1] != s[r+1] {
			break
		}
		l--
		r++
	}

	return s[l : r+1], r - l + 1
}

func longestPalindrome(s string) string {
	var ans string
	var ansLen int
	for mid := range s {
		l, r := mid, mid
		tmpS, tmpLen := getLongestPalindrome(s, l, r)
		if tmpLen > ansLen {
			ansLen = tmpLen
			ans = tmpS
		}

		l, r = mid, mid+1
		if s[l] == s[r] {
			tmpS, tmpLen = getLongestPalindrome(s, l, r)
			if tmpLen > ansLen {
				ansLen = tmpLen
				ans = tmpS
			}
		}
	}

	return ans
}

File: leetcode/cpp/5.cpp
#include "leetcode.h"

class Solution {
private:
	int getlongestPalindrome(const string &s, int &l, int &r)
	{
		int len = s.length();
		while (l-1 >= 0 && r+1 < len) {
			if (s[l-1] != s[r+1])
				break;
			--l;
			++r;
		}
		return r-l+1;
	}

	void updateAns(const string &s, const int l, const int r, string &ans, int &ansLen)
	{
		ans = s.substr(l, r-l+1);
		ansLen = r - l + 1;
	}

public:
	string longestPalindrome(string s)
	{
		int len = s.length();
		string ans;
		int ansLen = 0;
		for (int mid = 0; mid < len; ++mid) {
			
			int l = mid, r = mid;
			int tmpLen = getlongestPalindrome(s, l, r);
			if (tmpLen > ansLen)
				updateAns(s, l, r, ans, ansLen);

			l = mid, r = mid+1;
			if (r < len && s[l] == s[r]) {
				tmpLen = getlongestPalindrome(s, l, r);
				if (tmpLen > ansLen)
					updateAns(s, l, r, ans, ansLen);
			}
		}

		return ans;
	}
};